src/tasks/gitea/gitea-drone-oauth.ts

```diff
@@ -0,0 +1,225 @@
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      response_type: 'code',
+    },
+    maxRedirects: 1,
+    auth: {
+      username: env.GITEA_USER,
+      password: env.GITEA_PASSWORD,
+    },
+  }
+
+  console.log('Authorizing OAuth application')
+
+  const authorizeResponse: AxiosResponse<any> = await axios.get(`${giteaUrl}/login/oauth/authorize`, options)
+
+  return authorizeResponse.headers['set-cookie']
+}
+
+export function parseCSRFToken(authorizeHeaderCookies: string[]): string {
+  // Loop over cookies and find the _csrf cookie and retrieve the value
+  const csrfCookies = authorizeHeaderCookies.filter((cookie) => {
+    return cookie.includes(csrfCookieName) // Find cookie that contains _csrf
+  })
+
+  if (csrfCookies.length == 0) {
+    throw new GiteaDroneError('No CSRF cookie was returned')
+  }
+
+  const csrfTokens = csrfCookies[0]
+    .split(';') // Split that cookie into multiple parts, as a single cookie can contain multiple key/values ';' separated
+    .map((c) => c.trim())
+    .filter((cookie) => {
+      return cookie.includes(csrfCookieName) // Find the key/value pair with _csrf as key
+    })
+    .map((cookie) => {
+      return cookie.substring(csrfCookieName.length + 1) // Retrieve value for '_csrf'-key
+    })
+
+  if (csrfTokens.length == 0 || csrfTokens[0].length == 0) {
+    throw new GiteaDroneError('No CSRF token was returned')
+  }
+  return csrfTokens[0]
+}
+async function authorizeOAuthApp(giteaUrl: string, droneLoginUrl: string, oauthData: DroneSecret) {
+  let authorizeHeaderCookies: string[]
+  try {
+    authorizeHeaderCookies = await getGiteaAuthorizationHeaderCookie(giteaUrl, droneLoginUrl, oauthData)
+  } catch (error) {
+    throw new GiteaDroneError('Authorization already granted or something went wrong')
+  }
+
+  const csrfToken = parseCSRFToken(authorizeHeaderCookies)
+
+  console.log('Granting authorization')
+
+  const grantOptions: AxiosRequestConfig = {
+    method: 'POST',
+    url: `${giteaUrl}/login/oauth/grant`,
+    headers: {
+      cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+      Cookie: authorizeHeaderCookies
+        .map((cookie) => {
+          return cookie.split(';')[0]
+        })
+        .join('; '),
+    },
+    maxRedirects: 1,
+    auth: {
+      username: env.GITEA_USER,
+      password: env.GITEA_PASSWORD,
+    },
+    // Data for this post query must be stringified https://github.com/axios/axios#using-applicationx-www-form-urlencoded-format
+    data: querystring.stringify({
+      _csrf: csrfToken,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      client_id: `${oauthData.clientId}`,
+      // eslint-disable-next-line @typescript-eslint/camelcase
+      redirect_uri: droneLoginUrl,
+    }),
+  }
+
+  try {
+    await axios.request(grantOptions)
+  } catch (error) {
+    // Do nothing, error code could be on the redirect
```

Comment:
So you expect an error to occur but silence it? I would really like to see output of this so we can better handle this. Can you catch specific error codes if you expect them? And catch every one so that we run this to completion and have control instead of just wrapping it as a black box?

Reply:
Ok, that is fine. Just put these remarks there in the comment to be precise.